File: bindings/lua/tests/snapshot.lua
local dictionary = require("vinary_tree.libdictenstein")
local levenshtein = require("vinary_tree.liblevenshtein")

assert(levenshtein.status.ok == 0)
assert(levenshtein.status.end_of_stream == 1)
assert(levenshtein.status.domain_mismatch == 12)
local algorithms = {
  levenshtein.algorithm.standard,
  levenshtein.algorithm.transposition,
  levenshtein.algorithm.merge_and_split,
  levenshtein.algorithm.damerau_levenshtein,
}
local orders = {levenshtein.order.traversal, levenshtein.order.distance_then_term}
assert(#algorithms == 4 and #orders == 2)

local function collect(cursor)
  local output = {}
  for match in cursor do output[#output + 1] = {match.term, match.distance, match.id} end
  cursor:close()
  return output
end

for trace = 1, 64 do
  local words <close> = dictionary.dynamic_dawg("unicode")
  for index = 1, 16 do words:put("t" .. trace .. "-" .. index, index) end
  local automaton <close> = levenshtein.transducer(words)
  local expected = collect(automaton:query("", 64, "distance-then-term"))
  local cursor <close> = automaton:query("", 64, "distance-then-term")
  local actual = {cursor:next()}
  words:remove("t" .. trace .. "-1")
  words:put("t" .. trace .. "-2", 999)
  actual[#actual + 1] = cursor:next()
  words:clear()
  words:compact()
  words:put("after-" .. trace, 1000)
  for match in cursor do actual[#actual + 1] = match end
  assert(#actual == #expected)
  for index = 1, #actual do
    assert(actual[index].term == expected[index][1])
    assert(actual[index].distance == expected[index][2])
    assert(actual[index].id == expected[index][3])
  end
  words:close()
  local fresh = collect(automaton:query("", 64))
  assert(#fresh == 1 and fresh[1][1] == "after-" .. trace)
end

do
  local words <close> = dictionary.dynamic_dawg("unicode")
  words:put("cat", 1)
  words:put("cot", 2)
  local automaton <close> = levenshtein.transducer(words)
  local cache <close> = levenshtein.query_cache(automaton, 8, 1024 * 1024)
  local cold = collect(cache:query("cut", 1))
  local hit = collect(cache:query("cut", 1))
  assert(#cold == #hit)
  for index = 1, #cold do
    assert(cold[index][1] == hit[index][1] and cold[index][2] == hit[index][2])
  end
  local stats = cache:stats()
  assert(stats.requests == 2 and stats.hits == 1 and stats.misses == 1)
  assert(stats.resident_entries == 1 and stats.resident_weight > 0)
  assert(cache:reset_stats():stats().requests == 0)
  assert(cache:stats().resident_entries == 1)
  assert(cache:clear():stats().resident_entries == 0)
end

local dat <close> = dictionary.double_array_trie({{"café", 7}, {"caff"}}, "unicode")
assert(dat:get("café").value == 7)
assert(dat:get("caff").found and dat:get("caff").value == nil)

local suffixes <close> = dictionary.scdawg("unicode")
suffixes:put("cat", 1)
suffixes:put("cot", 2)
assert(suffixes:contains_substring("ot"))
assert(suffixes:frequency("t") == 2)
for _, algorithm in ipairs(algorithms) do
  local selected <close> = levenshtein.transducer(suffixes, algorithm)
  local exact = collect(selected:query("cat", 0, levenshtein.order.traversal))
  assert(#exact == 1 and exact[1][1] == "cat")
end

local bytes <close> = dictionary.dynamic_dawg("byte")
bytes:put("\0\255cat", 9)
local byte_automaton <close> = levenshtein.transducer(bytes)
assert(byte_automaton:domain() == "byte")
local byte_cursor <close> = byte_automaton:query_bytes("\0\255cot", 1)
local byte_batch = byte_cursor:next_batch(1)
assert(#byte_batch == 1 and byte_cursor:next_batch(1) == nil)
byte_cursor:close()
local byte_matches = {{byte_batch[1].term, byte_batch[1].distance, byte_batch[1].id}}
assert(#byte_matches == 1 and byte_matches[1][1] == "\0\255cat")
local reduction_cursor <close> = byte_automaton:query_bytes("\0\255cot", 1)
local reduced = reduction_cursor:reduce_batches(
  0,
  function(count, batch) return count + #batch end,
  1
)
assert(reduced == 1)
reduction_cursor:close()

local tokens <close> = dictionary.dynamic_dawg("u64")
assert(tokens:put_u64({1, 2, 3}, 12))
assert(tokens:contains_u64({1, 2, 3}))
assert(tokens:get_u64({1, 2, 3}).value == 12)
local token_automaton <close> = levenshtein.transducer(tokens)
assert(token_automaton:domain() == "u64")
local token_cursor <close> = token_automaton:query_u64({1, 2, 4}, 1)
local token_match = token_cursor:next()
assert(token_match.distance == 1 and token_match.term[3] == 3 and token_match.id == 12)
assert(tokens:remove_u64({1, 2, 3}))

local scratch = assert(os.getenv("VINARY_TREE_TEST_TMPDIR"),
  "VINARY_TREE_TEST_TMPDIR must name a writable non-tmpfs test directory")
local separator = package.config:sub(1, 1)
local function remove_persistence_artifacts(path)
  os.remove(path)
  os.remove(path .. ".wal")
  os.remove(path .. ".wlock")
end
local persistent_path = scratch .. separator .. "liblevenshtein-lua.artrie"
remove_persistence_artifacts(persistent_path)
do
  local persistent <close> = dictionary.create_persistent_artrie(persistent_path, "unicode")
  persistent:put("durable", 17)
  persistent:checkpoint()
end
do
  local persistent <close> = dictionary.open_persistent_artrie(persistent_path, "unicode")
  assert(persistent:get("durable").value == 17)
end

local vocabulary_path = scratch .. separator .. "liblevenshtein-lua.vocab"
remove_persistence_artifacts(vocabulary_path)
do
  local vocabulary <close> = dictionary.create_persistent_vocabulary(vocabulary_path)
  vocabulary:put("alpha", 0)
  vocabulary:checkpoint()
  assert(vocabulary:term(0) == "alpha")
end
do
  local vocabulary <close> = dictionary.open_persistent_vocabulary(vocabulary_path)
  assert(vocabulary:get("alpha").value == 0)
  assert(vocabulary:term(0) == "alpha")
end

local pattern <close> = levenshtein.phonetic_pattern("c[ao]t")
assert(pattern:matches("cat"))
assert(not pattern:matches("cut"))
local states, transitions = pattern:size()
assert(states > 0 and transitions > 0)
local llre <close> = levenshtein.llre_pattern("@name \"Greeting\"\n^hello$")
assert(llre:matches("hello") and not llre:matches("world"))
local product_automaton <close> = levenshtein.transducer(suffixes)
local product_pattern <close> = levenshtein.phonetic_pattern("c[ao]t")
local product_matches = collect(product_automaton:query_pattern(product_pattern, 0))
assert(#product_matches == 2)
local rules <close> = levenshtein.phonetic_rules(
  levenshtein.phonetic_rule_set_kind.english_orthography
)
assert(rules:len() > 0 and type(rules:apply("phone")) == "string")
local phonetic_rules <close> = levenshtein.phonetic_rules(
  levenshtein.phonetic_rule_set_kind.english_phonetic
)
assert(phonetic_rules:len() > 0)
assert(levenshtein.distance("kitten", "sitting") == 3)
assert(levenshtein.distance_threshold("kitten", "sitting", 3) == 3)
assert(levenshtein.damerau_distance("ca", "ac") == 1)
assert(levenshtein.damerau_distance_threshold("ca", "ac", 1) == 1)
assert(levenshtein.true_damerau_distance("ca", "ac") == 1)
assert(levenshtein.true_damerau_distance_threshold("ca", "ac", 1) == 1)
-- C6: distances count Unicode scalars, not bytes/UTF-16 code units.
assert(levenshtein.distance("café", "cafe") == 1)
assert(levenshtein.distance("🦀", "x") == 1)
assert(levenshtein.distance("é", "e") == 1)
-- C6: the exceeded-bound native sentinel (usize::MAX - 1) wraps to -2 under the
-- size_t -> lua_Integer conversion; "ca" -> "abc" separates OSA 3 from true 2.
assert(levenshtein.distance_threshold("kitten", "sitting", 2) == -2)
assert(levenshtein.damerau_distance_threshold("ca", "abc", 2) == -2)
assert(levenshtein.true_damerau_distance_threshold("ca", "abc", 2) == 2)
remove_persistence_artifacts(persistent_path)
remove_persistence_artifacts(vocabulary_path)
print("Lua binding snapshot integration passed")

File: bindings/ocaml/test/snapshot.ml
module Dict = Vinary_tree_libdictenstein
module Lev = Vinary_tree_liblevenshtein

let collect cursor =
  let values = List.of_seq (Lev.to_seq cursor) in
  Lev.cursor_close cursor;
  values

let text_of_match result =
  match result.Lev.term with
  | Lev.Text text -> text
  | Lev.Tokens _ -> failwith "expected a text match"

let remove_if_present path =
  if Sys.file_exists path then Sys.remove path

let scratch_directory =
  match Sys.getenv_opt "VINARY_TREE_TEST_TMPDIR" with
  | Some path -> path
  | None ->
      invalid_arg
        "VINARY_TREE_TEST_TMPDIR must name a writable non-tmpfs test directory"

let temporary_path suffix =
  let path = Filename.concat scratch_directory ("liblevenshtein-ocaml" ^ suffix) in
  List.iter remove_if_present [path; path ^ ".wal"; path ^ ".wlock"];
  path

let () =
  let algorithms : Lev.algorithm list =
    [Lev.Standard; Lev.Transposition; Lev.Merge_and_split;
     Lev.Damerau_levenshtein]
  in
  let orders : Lev.query_order list = [Lev.Traversal; Lev.Distance_then_term] in
  assert (List.length algorithms = 4 && List.length orders = 2);

  for trace = 0 to 63 do
    let words = Dict.dynamic_dawg () in
    let entries =
      Array.init 16 (fun index ->
        (Printf.sprintf "t%d-%d" trace index, Some (Int64.of_int index)))
    in
    assert (Dict.put_many words entries = 16);
    let automaton = Lev.transducer (Dict.resource words) in
    let expected =
      collect (Lev.query ~order:Lev.Distance_then_term automaton ""
                 ~maximum_distance:64)
    in
    let cursor =
      Lev.query ~order:Lev.Distance_then_term automaton "" ~maximum_distance:64
    in
    let first = Option.get (Lev.next cursor) in
    assert (Dict.remove words (Printf.sprintf "t%d-1" trace));
    ignore (Dict.put words (Printf.sprintf "t%d-2" trace) (Some 999L));
    let second = Option.get (Lev.next cursor) in
    Dict.clear words;
    ignore (Dict.compact words);
    ignore (Dict.put words (Printf.sprintf "after-%d" trace) (Some 1000L));
    let remainder = List.of_seq (Lev.to_seq cursor) in
    Lev.cursor_close cursor;
    assert (first :: second :: remainder = expected);
    Dict.close words;
    let fresh = collect (Lev.query automaton "" ~maximum_distance:64) in
    assert (List.map text_of_match fresh = [Printf.sprintf "after-%d" trace]);
    Lev.close_transducer automaton
  done;

  let dat = Dict.double_array_trie [|("café", Some 7L); ("caff", None)|] in
  assert ((Dict.get dat "café").value = Some 7L);
  assert ((Dict.get dat "caff").found);
  Dict.close dat;

  let suffixes = Dict.scdawg () in
  ignore (Dict.put suffixes "cat" (Some 1L));
  ignore (Dict.put suffixes "cot" (Some 2L));
  assert (Dict.contains_substring suffixes "ot");
  assert (Dict.substring_frequency suffixes "t" = 2);
  Dict.close suffixes;

  let token_dictionary = Dict.dynamic_dawg ~domain:Vinary_tree_interop.U64 () in
  assert (Dict.put_u64 token_dictionary [|1L; 2L; 3L|] (Some 12L));
  let token_automaton = Lev.transducer (Dict.resource token_dictionary) in
  let token_cursor =
    Lev.query_u64 token_automaton [|1L; 2L; 4L|] ~maximum_distance:1
  in
  let token_match = Option.get (Lev.next token_cursor) in
  (match token_match.term with
   | Lev.Tokens tokens -> assert (tokens = [|1L; 2L; 3L|])
   | Lev.Text _ -> assert false);
  assert (token_match.distance = 1 && token_match.id = Some 12L);
  Lev.cursor_close token_cursor;
  Dict.close token_dictionary;
  Lev.close_transducer token_automaton;

  let byte_dictionary =
    Dict.dynamic_dawg ~domain:Vinary_tree_interop.Byte ()
  in
  let byte_text = "\x63\x00\xff" in
  assert (Dict.put byte_dictionary byte_text (Some 13L));
  let byte_automaton = Lev.transducer (Dict.resource byte_dictionary) in
  let byte_cursor =
    Lev.query_bytes byte_automaton (Bytes.of_string byte_text)
      ~maximum_distance:0
  in
  let byte_batch = Option.get (Lev.next_batch ~maximum:1 byte_cursor) in
  assert (Array.length byte_batch = 1);
  (match byte_batch.(0).term with
   | Lev.Text text -> assert (text = byte_text)
   | Lev.Tokens _ -> assert false);
  assert (byte_batch.(0).distance = 0 && byte_batch.(0).id = Some 13L);
  assert (Lev.next_batch ~maximum:1 byte_cursor = None);
  Lev.cursor_close byte_cursor;
  Dict.close byte_dictionary;
  Lev.close_transducer byte_automaton;

  let persistent_path = temporary_path ".artrie" in
  let persistent = Dict.create_persistent_artrie persistent_path in
  ignore (Dict.put persistent "durable" (Some 17L));
  Dict.checkpoint persistent;
  Dict.close persistent;
  let reopened = Dict.open_persistent_artrie persistent_path in
  assert ((Dict.get reopened "durable").value = Some 17L);
  Dict.close reopened;

  let vocabulary_path = temporary_path ".vocab" in
  let vocabulary = Dict.create_persistent_vocabulary vocabulary_path in
  ignore (Dict.put vocabulary "alpha" (Some 0L));
  Dict.checkpoint vocabulary;
  assert (Dict.term vocabulary 0L = Some "alpha");
  Dict.close vocabulary;
  let reopened_vocabulary = Dict.open_persistent_vocabulary vocabulary_path in
  assert ((Dict.get reopened_vocabulary "alpha").value = Some 0L);
  Dict.close reopened_vocabulary;

  let pattern = Lev.regex_pattern "c[ao]t" in
  assert (Lev.pattern_matches pattern "cat");
  assert (not (Lev.pattern_matches pattern "cut"));
  let states, transitions = Lev.pattern_size pattern in
  assert (states > 0 && transitions > 0);
  Lev.close_pattern pattern;
  let rules = Lev.phonetic_rules "english-orthography" in
  assert (Lev.rules_length rules > 0);
  ignore (Lev.apply_rules rules "phone");
  Lev.close_rules rules;
  assert (Lev.distance "kitten" "sitting" = 3);
  assert (Lev.distance_threshold "kitten" "sitting" 3 = 3);
  assert (Lev.damerau_distance "ca" "ac" = 1);
  assert (Lev.damerau_distance_threshold "ca" "ac" 1 = 1);
  assert (Lev.true_damerau_distance "ca" "ac" = 1);
  assert (Lev.true_damerau_distance_threshold "ca" "ac" 1 = 1);

  (* C5 (reduced): the pull iterator (to_seq) and the push reducer (fold_batches)
     drain the same query to the same multiset of matched terms. *)
  let equality_dictionary = Dict.dynamic_dawg () in
  ignore (Dict.put equality_dictionary "cat" (Some 1L));
  ignore (Dict.put equality_dictionary "cot" (Some 2L));
  ignore (Dict.put equality_dictionary "cut" (Some 3L));
  let equality_automaton = Lev.transducer (Dict.resource equality_dictionary) in
  List.iter
    (fun algorithm ->
       let selected =
         Lev.transducer ~algorithm (Dict.resource equality_dictionary)
       in
       let exact =
         collect
           (Lev.query ~order:Lev.Traversal selected "cat" ~maximum_distance:0)
       in
       assert (List.map text_of_match exact = ["cat"]);
       Lev.close_transducer selected)
    algorithms;
  let by_iterator =
    List.sort compare
      (List.map (fun result -> result.Lev.term)
         (collect (Lev.query equality_automaton "cat" ~maximum_distance:1)))
  in
  let reducer_cursor = Lev.query equality_automaton "cat" ~maximum_distance:1 in
  let by_reducer =
    List.sort compare
      (Lev.fold_batches reducer_cursor []
         (fun state batch ->
            Array.fold_left (fun acc result -> result.Lev.term :: acc) state batch))
  in
  Lev.cursor_close reducer_cursor;
  assert (by_iterator = by_reducer);
  assert (List.length by_iterator = 3);
  let llre = Lev.llre_pattern "@name \"Greeting\"\n^hello$" in
  assert (Lev.pattern_matches llre "hello");
  assert (not (Lev.pattern_matches llre "world"));
  Lev.close_pattern llre;
  let product_pattern = Lev.regex_pattern "c[ao]t" in
  let product_matches =
    collect
      (Lev.query_pattern equality_automaton product_pattern ~maximum_distance:0)
  in
  assert
    (List.sort compare (List.map text_of_match product_matches) = ["cat"; "cot"]);
  Lev.close_pattern product_pattern;
  Dict.close equality_dictionary;
  Lev.close_transducer equality_automaton;

  (* C6 (reduced): distances decode UTF-8 to Unicode scalar values, so a
     multi-byte character is one edit, not one edit per byte. *)
  assert (Lev.distance "café" "cafe" = 1);
  assert (Lev.distance "🦀" "x" = 1);
  assert (Lev.distance "é" "e" = 1);
  assert (Lev.distance_threshold "café" "cafe" 1 = 1);

  (* C2/C3 (reduced): close is idempotent and a closed pattern raises
     Invalid_argument on use rather than dereferencing a freed automaton. *)
  let lifecycle = Lev.regex_pattern "c[ao]t" in
  assert (Lev.pattern_matches lifecycle "cat");
  Lev.close_pattern lifecycle;
  Lev.close_pattern lifecycle;
  (match Lev.pattern_matches lifecycle "cat" with
   | (_ : bool) -> assert false
   | exception Invalid_argument _ -> ());

  List.iter remove_if_present
    [ persistent_path; persistent_path ^ ".wal"; persistent_path ^ ".wlock";
      vocabulary_path; vocabulary_path ^ ".wal"; vocabulary_path ^ ".wlock" ];
  print_endline "OCaml binding snapshot integration passed"
